shared/actions/signup/index.js

```diff
@@ -156,219 +83,344 @@ function isValidName (name: ?string): ?string {
-      incomingCallMap: {},
-      callback: err => {
-        if (err) {
-          console.error("username isn't available:", err)
-          dispatch({
-            type: Constants.checkUsernameEmail,
-            error: true,
-            payload: {emailError, usernameError: `Username error: ${err.message}`, email, username}
-          })
-          resolve()
-        } else {
-          // We need this check to make flow happy. This should never be null
-          if (username && email) {
-            dispatch({
-              type: Constants.checkUsernameEmail,
-              payload: {username, email}
-            })
-            dispatch(nextPhase())
-            resolve()
-          } else {
-            reject()
-          }
+// This one is a bit different, since we don't want to change how everything in engine works just yet.
+// It's still a thunked actions
+function signupWithService (email: string, username: string, inviteCode: string, passphrase: HiddenString, deviceName: string, skipMail: boolean) {
+  return (dispatch, getState) => {
+    const params: signup_signup_rpc = {
+      method: 'signup.signup',
+      waitingHandler: () => {},
+      param: {
+        email,
+        inviteCode,
+        username,
+        deviceName,
+        passphrase: passphrase.stringValue(),
+        storeSecret: false,
+        skipMail
+      },
+      incomingCallMap: {
+        'keybase.1.loginUi.displayPrimaryPaperKey': ({sessionID, phrase}, response) => {
+          dispatch({type: 'keybase.1.loginUi.displayPrimaryPaperKey', payload: {phrase, response}})
+        },
+        'keybase.1.gpgUi.wantToAddGPGKey': (params, {error, result}) => {
+          // Do not add a gpg key for now
+          result(false)
         }
+      },
+      callback: (error, result) => {
+        dispatch({type: 'keybase.1.signup.signup', payload: {error, result}})
       }
     }
-
     engine.rpc(params)
-  })
+  }
 }
 
-export function checkPassphrase (passphrase1: string, passphrase2: string): TypedAsyncAction<CheckPassphrase | RouteAppend> {
-  return dispatch => new Promise((resolve, reject) => {
-    let passphraseError = null
-    if (!passphrase1 || !passphrase2) {
-      passphraseError = new HiddenString('Fields cannot be blank')
-    } else if (passphrase1 !== passphrase2) {
-      passphraseError = new HiddenString('Passphrases must match')
-    } else if (passphrase1.length < 12) {
-      passphraseError = new HiddenString('Passphrase must be at least 12 Characters')
+function requestInviteWithService (email: string, name: string, callback: NodeCB) {
+  const params: signup_inviteRequest_rpc = {
+    method: 'signup.inviteRequest',
+    param: {
+      email: email,
+      fullname: name,
+      notes: 'Requested through GUI app'
+    },
+    waitingHandler: () => {},
+    incomingCallMap: {},
+    callback
+  }
+  engine.rpc(params)
+}
+
+// Sagas
+
+function * requestInviteSaga () {
+  while (true) {
```

Comment:
These `while true`'s are because the user can stay on this screen and type in something with an error so the user and us have to start the flow all over again.